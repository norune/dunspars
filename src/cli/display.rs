use std::fmt;
use std::io::{stdout, Write};

use anyhow::Result;
use indoc::writedoc;
use owo_colors::{colors::xterm, Style};

use crate::pokemon::{self, Move, MoveList, Pokemon, Stats, TypeChart};

pub struct StyleSheet {
    pub default: Style,
    pub header: Style,
    pub quad: Style,
    pub double: Style,
    pub neutral: Style,
    pub half: Style,
    pub quarter: Style,
    pub zero: Style,
    pub power: Style,
    pub accuracy: Style,
    pub pp: Style,
    pub hp: Style,
    pub attack: Style,
    pub defense: Style,
    pub special_attack: Style,
    pub special_defense: Style,
    pub speed: Style,
    pub move_: Style,
}

impl Default for StyleSheet {
    fn default() -> Self {
        Self {
            default: Style::new().default_color(),
            header: Style::new().bright_green().bold(),
            quad: Style::new().red(),
            double: Style::new().yellow(),
            neutral: Style::new().green(),
            half: Style::new().blue(),
            quarter: Style::new().bright_cyan(),
            zero: Style::new().purple(),
            power: Style::new().fg::<xterm::FlushOrange>(),
            accuracy: Style::new().fg::<xterm::FernGreen>(),
            pp: Style::new().fg::<xterm::ScienceBlue>(),
            hp: Style::new().fg::<xterm::GuardsmanRed>(),
            attack: Style::new().fg::<xterm::DecoOrange>(),
            defense: Style::new().fg::<xterm::AeroBlue>(),
            special_attack: Style::new().fg::<xterm::BlazeOrange>(),
            special_defense: Style::new().fg::<xterm::PoloBlue>(),
            speed: Style::new().fg::<xterm::PurplePizzazz>(),
            move_: Style::new().fg::<xterm::SpringGreen>(),
        }
    }
}

pub struct PokemonDisplay<'a, 'b> {
    pokemon: &'a Pokemon<'b>,
    css: StyleSheet,
}

impl<'a, 'b> PokemonDisplay<'a, 'b> {
    pub fn new(pokemon: &'a Pokemon<'b>) -> Self {
        Self {
            pokemon,
            css: StyleSheet::default(),
        }
    }

    pub fn print(&self) -> Result<()> {
        let Pokemon {
            name,
            version,
            generation,
            primary_type,
            secondary_type,
            stats,
            ..
        } = self.pokemon;

        let stats_display = StatsDisplay::new(stats);

        let mut f = stdout().lock();
        writedoc! {
            f,
            "

            {name} {version}({generation})
            {primary_type} {secondary_type}
            {stats_display}
            ",
            name = self.css.header.style(name),
            secondary_type = secondary_type.as_ref().unwrap_or(&"".to_string())
        }?;
        Ok(())
    }
}

pub struct StatsDisplay<'a> {
    stats: &'a Stats,
    css: StyleSheet,
}

impl<'a> StatsDisplay<'a> {
    pub fn new(stats: &'a Stats) -> Self {
        Self {
            stats,
            css: StyleSheet::default(),
        }
    }
}

impl fmt::Display for StatsDisplay<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let Stats {
            hp,
            attack,
            defense,
            special_attack,
            special_defense,
            speed,
        } = self.stats;

        writedoc! {
            f,
            "
            hp    atk   def   satk  sdef  spd
            {hp:<6}{attack:<6}{defense:<6}{special_attack:<6}{special_defense:<6}{speed:<6}            
            ",
            hp = self.css.hp.style(hp),
            attack = self.css.attack.style(attack),
            defense = self.css.defense.style(defense),
            special_attack = self.css.special_attack.style(special_attack),
            special_defense = self.css.special_defense.style(special_defense),
            speed = self.css.speed.style(speed),
        }
    }
}

pub struct TypeChartDisplay<'a> {
    type_chart: &'a TypeChart,
    label: &'static str,
    css: StyleSheet,
}

impl<'a> TypeChartDisplay<'a> {
    pub fn new(type_chart: &'a TypeChart, label: &'static str) -> Self {
        TypeChartDisplay {
            type_chart,
            label,
            css: StyleSheet::default(),
        }
    }

    pub fn print(&self) -> Result<()> {
        let weakness_groups = WeaknessGroups::new(self.type_chart.get_value(), |t| {
            Some((t.0.clone(), t.1.clone()))
        });

        println!("\n{}", self.css.header.style(self.label));
        self.print_weakness_groups(weakness_groups)
    }

    fn print_weakness_groups(&self, weakness_groups: WeaknessGroups<String>) -> Result<()> {
        let mut f = stdout().lock();

        if weakness_groups.quad.len() > 0 {
            let quad = weakness_groups.quad.join(" ");
            writeln!(f, "quad:\n{}", self.css.quad.style(quad))?;
        }
        if weakness_groups.double.len() > 0 {
            let double = weakness_groups.double.join(" ");
            writeln!(f, "double:\n{}", self.css.double.style(double))?;
        }
        if weakness_groups.neutral.len() > 0 {
            let neutral = weakness_groups.neutral.join(" ");
            writeln!(f, "neutral:\n{}", self.css.neutral.style(neutral))?;
        }
        if weakness_groups.half.len() > 0 {
            let half = weakness_groups.half.join(" ");
            writeln!(f, "half:\n{}", self.css.half.style(half))?;
        }
        if weakness_groups.quarter.len() > 0 {
            let quarter = weakness_groups.quarter.join(" ");
            writeln!(f, "quarter:\n{}", self.css.quarter.style(quarter))?;
        }
        if weakness_groups.zero.len() > 0 {
            let zero = weakness_groups.zero.join(" ");
            writeln!(f, "zero:\n{}", self.css.zero.style(zero))?;
        }
        if weakness_groups.other.len() > 0 {
            let other = weakness_groups.other.join(" ");
            writeln!(f, "zero:\n{}", other)?;
        }

        Ok(())
    }
}

pub struct MatchDisplay<'a, 'b, 'c, 'd> {
    type_chart: &'a TypeChart,
    move_list: &'b MoveList<'d>,
    defender: &'c Pokemon<'d>,
    attacker: &'c Pokemon<'d>,
    stab_only: bool,
    css: StyleSheet,
}

impl<'a, 'b, 'c, 'd> MatchDisplay<'a, 'b, 'c, 'd> {
    pub fn new(
        type_chart: &'a TypeChart,
        move_list: &'b MoveList<'d>,
        defender: &'c Pokemon<'d>,
        attacker: &'c Pokemon<'d>,
        stab_only: bool,
    ) -> Self {
        MatchDisplay {
            type_chart,
            move_list,
            defender,
            attacker,
            stab_only,
            css: StyleSheet::default(),
        }
    }

    pub fn print(&self) -> Result<()> {
        self.print_stats()?;

        let weakness_groups = WeaknessGroups::new(self.move_list.get_value(), |move_| {
            let stab_qualified = !self.stab_only || pokemon::is_stab(&move_.1.type_, self.attacker);
            if move_.1.damage_class != "status" && stab_qualified {
                let multiplier = self.type_chart.get_multiplier(&move_.1.type_);
                Some((move_.1, multiplier))
            } else {
                None
            }
        });

        let header = format!("{}'s moves vs {}", self.attacker.name, self.defender.name);
        println!("\n{}", self.css.header.style(header));

        self.print_weakness_groups(weakness_groups)
    }

    fn print_stats(&self) -> Result<()> {
        let defender_stats = StatsDisplay::new(&self.defender.stats);
        let attacker_stats = StatsDisplay::new(&self.attacker.stats);

        let mut f = stdout().lock();
        let defender_header = format!("{}'s stats", self.defender.name);
        let attacker_header = format!("{}'s stats", self.attacker.name);
        writedoc! {
            f,
            "
            {defender_header}
            {defender_stats}
            {attacker_header}
            {attacker_stats}
            ",
            defender_header = self.css.header.style(defender_header),
            attacker_header = self.css.header.style(attacker_header),
        }?;

        Ok(())
    }

    fn print_weakness_groups(&self, weakness_groups: WeaknessGroups<&Move>) -> Result<()> {
        let mut f = stdout().lock();

        if weakness_groups.quad.len() > 0 {
            write!(f, "quad: ")?;
            self.print_group(weakness_groups.quad, self.css.quad)?;
        }
        if weakness_groups.double.len() > 0 {
            write!(f, "double: ")?;
            self.print_group(weakness_groups.double, self.css.double)?;
        }
        if weakness_groups.neutral.len() > 0 {
            write!(f, "neutral: ")?;
            self.print_group(weakness_groups.neutral, self.css.neutral)?;
        }
        if weakness_groups.half.len() > 0 {
            write!(f, "half: ")?;
            self.print_group(weakness_groups.half, self.css.half)?;
        }
        if weakness_groups.quarter.len() > 0 {
            write!(f, "quarter: ")?;
            self.print_group(weakness_groups.quarter, self.css.quarter)?;
        }
        if weakness_groups.zero.len() > 0 {
            write!(f, "zero: ")?;
            self.print_group(weakness_groups.zero, self.css.zero)?;
        }
        if weakness_groups.other.len() > 0 {
            write!(f, "other: ")?;
            self.print_group(weakness_groups.other, self.css.neutral)?;
        }

        Ok(())
    }

    fn print_group(&self, group: Vec<&Move>, group_style: Style) -> Result<()> {
        let mut f = stdout().lock();
        for move_ in group {
            let damage_class = match move_.damage_class.as_str() {
                "special" => "s",
                "physical" => "p",
                _ => "?",
            };
            let move_string = format!("{}({})", move_.name, damage_class);
            let styled_move;
            if pokemon::is_stab(&move_.type_, self.attacker) {
                styled_move = group_style.underline().style(move_string);
            } else {
                styled_move = group_style.style(move_string);
            }

            write!(f, "{} ", styled_move)?;
        }
        writeln!(f, "{}", "\n")?;
        Ok(())
    }
}

pub struct WeaknessGroups<T> {
    pub quad: Vec<T>,
    pub double: Vec<T>,
    pub neutral: Vec<T>,
    pub half: Vec<T>,
    pub quarter: Vec<T>,
    pub zero: Vec<T>,
    pub other: Vec<T>,
}

impl<T> WeaknessGroups<T> {
    pub fn new<C, F, I>(collection: C, mut cb: F) -> Self
    where
        C: IntoIterator<Item = I>,
        F: FnMut(I) -> Option<(T, f32)>,
    {
        let mut groups = WeaknessGroups {
            quad: vec![],
            double: vec![],
            neutral: vec![],
            half: vec![],
            quarter: vec![],
            zero: vec![],
            other: vec![],
        };

        for element in collection {
            if let Some(result) = cb(element) {
                let (item, multiplier) = result;
                match multiplier {
                    x if x == 4.0 => groups.quad.push(item),
                    x if x == 2.0 => groups.double.push(item),
                    x if x == 1.0 => groups.neutral.push(item),
                    x if x == 0.5 => groups.half.push(item),
                    x if x == 0.25 => groups.quarter.push(item),
                    x if x == 0.0 => groups.zero.push(item),
                    _ => groups.other.push(item),
                }
            }
        }

        groups
    }
}

pub struct MoveListDisplay<'a, 'b, 'c> {
    move_list: &'a MoveList<'c>,
    pokemon: &'b Pokemon<'c>,
    css: StyleSheet,
}

impl<'a, 'b, 'c> MoveListDisplay<'a, 'b, 'c> {
    pub fn new(move_list: &'a MoveList<'c>, pokemon: &'b Pokemon<'c>) -> Self {
        MoveListDisplay {
            move_list,
            pokemon,
            css: StyleSheet::default(),
        }
    }

    pub fn print(&self) -> Result<()> {
        println!("\n{}", self.css.header.style("moves"));

        let mut f = stdout().lock();
        for move_ in self.move_list.get_value() {
            let Move {
                name,
                accuracy,
                power,
                pp,
                damage_class,
                type_,
                ..
            } = move_.1;

            let is_stab = pokemon::is_stab(&move_.1.type_, self.pokemon);
            let stab = if is_stab { "(s)" } else { "" };

            let move_name = format!("{name}{stab}", name = self.css.move_.style(name));
            let type_damage = format!("{type_} {damage_class}");
            let stats = format!(
                "power: {:3}  accuracy: {:3}  pp: {:2}",
                self.css.power.style(power.unwrap_or(0)),
                self.css.accuracy.style(accuracy.unwrap_or(0)),
                self.css.pp.style(pp.unwrap_or(0))
            );

            let default_learn = ("".to_string(), 0i64);
            let (learn_method, level) = self.pokemon.moves.get(move_.0).unwrap_or(&default_learn);
            let learn_level = if *level == 0i64 {
                "".to_string()
            } else {
                level.to_string()
            };
            let learn = format!("{} {}", learn_method, learn_level);

            writedoc! {
                f,
                "
                {move_name:33}{type_damage:22}{stats:80}{learn}
                ",
            }?;
        }

        Ok(())
    }
}

pub struct MoveDisplay<'a, 'b> {
    move_: &'a Move<'b>,
    css: StyleSheet,
}

impl<'a, 'b> MoveDisplay<'a, 'b> {
    pub fn new(move_: &'a Move<'b>) -> Self {
        MoveDisplay {
            move_,
            css: StyleSheet::default(),
        }
    }

    pub fn print(&self) -> Result<()> {
        let Move {
            power,
            accuracy,
            pp,
            name,
            effect,
            damage_class,
            type_,
            ..
        } = self.move_;

        let stats = format!(
            "power: {:3}  accuracy: {:3}  pp: {:3}",
            self.css.power.style(power.unwrap_or(0)),
            self.css.accuracy.style(accuracy.unwrap_or(0)),
            self.css.pp.style(pp.unwrap_or(0))
        );

        let mut f = stdout().lock();
        writedoc! {
            f,
            "

            {name}
            {type_} {damage_class}
            {stats}
            {effect}

            ",
            name = self.css.header.style(name)
        }?;

        Ok(())
    }
}
