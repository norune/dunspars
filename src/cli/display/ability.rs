use super::{Colors, DisplayComponent};
use crate::data::Ability;

use std::fmt;

use indoc::writedoc;

impl fmt::Display for DisplayComponent<&Ability> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let Ability { name, effect, .. } = self.context;

        writedoc! {
            f,
            "{header}{name}{header:#}
            {effect}",
            header = self.ansi_bold(Colors::Header)
        }
    }
}
