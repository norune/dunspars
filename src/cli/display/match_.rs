use super::{Colors, DisplayComponent, MoveWeaknessComponent};
use crate::models::Pokemon;

use std::fmt;

use indoc::writedoc;
use rusqlite::Connection;

pub struct MatchComponent<'a> {
    pub defender: &'a Pokemon,
    pub attacker: &'a Pokemon,
    pub db: &'a Connection,
    pub verbose: bool,
    pub stab_only: bool,
}

impl fmt::Display for DisplayComponent<MatchComponent<'_>> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let MatchComponent {
            defender,
            attacker,
            db,
            verbose,
            stab_only,
        } = self.context;

        let defender_stats = DisplayComponent::new(&defender.stats, self.color_enabled);
        let attacker_stats = DisplayComponent::new(&attacker.stats, self.color_enabled);

        let defender_moves_header = format!("{}'s moves vs {}", attacker.name, defender.name);
        let defender_context = MoveWeaknessComponent {
            defender,
            attacker,
            db,
            verbose,
            stab_only,
        };
        let defender_weaknesses = DisplayComponent::new(defender_context, self.color_enabled);

        let attacker_moves_header = format!("{}'s moves vs {}", defender.name, attacker.name);
        let attacker_context = MoveWeaknessComponent {
            defender: attacker,
            attacker: defender,
            db,
            verbose,
            stab_only,
        };
        let attacker_weaknesses = DisplayComponent::new(attacker_context, self.color_enabled);

        writedoc! {
            f,
            "{header}{defender_header}{header:#} {defender_primary_type} {defender_secondary_type}
            {defender_stats}
            {header}{attacker_header}{header:#} {attacker_primary_type} {attacker_secondary_type}
            {attacker_stats}

            {header}{defender_moves_header}{header:#}{defender_weaknesses}

            {header}{attacker_moves_header}{header:#}{attacker_weaknesses}",
            defender_header = &defender.name,
            defender_primary_type = defender.primary_type,
            defender_secondary_type = defender.secondary_type.as_deref().unwrap_or(""),
            attacker_header = &attacker.name,
            attacker_primary_type = attacker.primary_type,
            attacker_secondary_type = attacker.secondary_type.as_deref().unwrap_or(""),
            header = self.ansi_bold(Colors::Header),
        }
    }
}
