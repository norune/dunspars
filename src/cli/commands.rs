use super::display::*;
use super::ResourceArgs;
use crate::api::game_to_gen;
use crate::models::resource::{AbilityRow, GameRow, MoveRow, PokemonRow, Resource, TypeRow};
use crate::models::{Ability, Move, Pokemon, PokemonData, Type};
use crate::resource::config::{Config, ConfigFile};
use crate::resource::database::DatabaseFile;

use std::io::Write;

use anyhow::{anyhow, Result};
use indoc::writedoc;
use rusqlite::Connection;

struct DbContext {
    db: Connection,
    config: Config,
}
impl DbContext {
    fn try_new(config: Config) -> Result<Self> {
        let file = DatabaseFile::default();
        let db = file.connect()?;

        Ok(Self { db, config })
    }

    fn get_generation(&self) -> Result<u8> {
        let game = match &self.config.game {
            Some(game) => self.validate::<GameRow>(game)?,
            None => self
                .get_latest_game()
                .ok_or(anyhow!("Cannot find the latest game"))?,
        };
        Ok(game_to_gen(&game, &self.db))
    }

    fn validate<T: Resource>(&self, name: &str) -> Result<String> {
        T::validate(name, &self.db)
    }

    fn get_latest_game(&self) -> Option<String> {
        GameRow::resource(&self.db).last().map(|g| g.to_string())
    }
}

pub trait Command {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32>;
}

pub struct SetupCommand;
impl Command for SetupCommand {
    async fn run(&self, _config: Config, writer: &mut impl Write) -> Result<i32> {
        let file = DatabaseFile::default();
        file.build_db(writer).await?;
        Ok(0)
    }
}

pub struct PokemonCommand {
    pub name: String,
    pub moves: bool,
    pub evolution: bool,
}
impl Command for PokemonCommand {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32> {
        let ctx = DbContext::try_new(config)?;
        let generation = ctx.get_generation()?;

        let pokemon_name = ctx.validate::<PokemonRow>(&self.name)?;
        let pokemon = PokemonData::from_name(&pokemon_name, generation, &ctx.db)?;
        let pokemon_display = DisplayComponent::new(&pokemon, ctx.config.color_enabled);

        let defense_chart = pokemon.get_defense_chart(&ctx.db)?;
        let defense_chart_ctx = TypeChartComponent {
            type_chart: &defense_chart,
        };
        let type_chart_display = DisplayComponent::new(defense_chart_ctx, ctx.config.color_enabled);

        writedoc! {
            writer,
            "
            {pokemon_display}

            {type_chart_display}
            "
        }?;

        if self.evolution {
            let evolution_step = pokemon.get_evolution_steps(&ctx.db)?;
            let evolution_step_display =
                DisplayComponent::new(&evolution_step, ctx.config.color_enabled);
            writedoc! {
                writer,
                "

                {evolution_step_display}
                "
            }?;
        }

        if self.moves {
            let moves = pokemon.get_moves(&ctx.db)?;
            let move_list_context = MoveListComponent {
                move_list: &moves,
                pokemon: &pokemon,
            };
            let move_list_display =
                DisplayComponent::new(move_list_context, ctx.config.color_enabled);

            writedoc! {
                writer,
                "

                {move_list_display}
                "
            }?;
        }

        Ok(0)
    }
}

pub struct TypeCommand {
    pub primary_type: String,
    pub secondary_type: Option<String>,
}
impl TypeCommand {
    fn get_type(type_name: &str, generation: u8, ctx: &DbContext) -> Result<Type> {
        let type_name = ctx.validate::<TypeRow>(type_name)?;
        let type_ = Type::from_name(&type_name, generation, &ctx.db)?;
        Ok(type_)
    }
}
impl Command for TypeCommand {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32> {
        let ctx = DbContext::try_new(config)?;
        let generation = ctx.get_generation()?;

        let primary_type = Self::get_type(&self.primary_type, generation, &ctx)?;
        let primary_offense_ctx = TypeChartComponent {
            type_chart: &primary_type.offense_chart,
        };
        let primary_offense_display =
            DisplayComponent::new(primary_offense_ctx, ctx.config.color_enabled);

        let secondary_type = self
            .secondary_type
            .as_ref()
            .map(|t| Self::get_type(t, generation, &ctx));

        match secondary_type {
            Some(secondary_type) => {
                let secondary_type = secondary_type?;
                let secondary_offense_ctx = TypeChartComponent {
                    type_chart: &secondary_type.offense_chart,
                };
                let secondary_offense_display =
                    DisplayComponent::new(secondary_offense_ctx, ctx.config.color_enabled);

                let combined_defense = primary_type.defense_chart + secondary_type.defense_chart;
                let defense_ctx = TypeChartComponent {
                    type_chart: &combined_defense,
                };
                let defense_display = DisplayComponent::new(defense_ctx, ctx.config.color_enabled);

                writedoc! {
                    writer,
                    "
                    {primary_offense_display}

                    {secondary_offense_display}

                    {defense_display}
                    "
                }?;
            }
            None => {
                let defense_ctx = TypeChartComponent {
                    type_chart: &primary_type.defense_chart,
                };
                let defense_display = DisplayComponent::new(defense_ctx, ctx.config.color_enabled);

                writedoc! {
                    writer,
                    "
                    {primary_offense_display}

                    {defense_display}
                    "
                }?;
            }
        }

        Ok(0)
    }
}

pub struct MoveCommand {
    pub name: String,
}
impl Command for MoveCommand {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32> {
        let ctx = DbContext::try_new(config)?;
        let generation = ctx.get_generation()?;

        let move_name = ctx.validate::<MoveRow>(&self.name)?;
        let move_ = Move::from_name(&move_name, generation, &ctx.db)?;
        let move_display = DisplayComponent::new(&move_, ctx.config.color_enabled);

        writedoc! {
            writer,
            "
            {move_display}
            "
        }?;

        Ok(0)
    }
}

pub struct AbilityCommand {
    pub name: String,
}
impl Command for AbilityCommand {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32> {
        let ctx = DbContext::try_new(config)?;
        let generation = ctx.get_generation()?;

        let ability_name = ctx.validate::<AbilityRow>(&self.name)?;
        let ability = Ability::from_name(&ability_name, generation, &ctx.db)?;
        let ability_display = DisplayComponent::new(&ability, ctx.config.color_enabled);

        writedoc! {
            writer,
            "
            {ability_display}
            "
        }?;

        Ok(0)
    }
}

#[derive(Clone)]
pub struct MatchCommand {
    pub defender_names: Vec<String>,
    pub attacker_name: String,
    pub verbose: bool,
    pub stab_only: bool,
}
impl Command for MatchCommand {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32> {
        let ctx = DbContext::try_new(config)?;
        let generation = ctx.get_generation()?;

        let attacker_name = ctx.validate::<PokemonRow>(&self.attacker_name)?;
        let attacker_data = PokemonData::from_name(&attacker_name, generation, &ctx.db)?;
        let attacker_moves = attacker_data.get_moves(&ctx.db)?;
        let attacker_chart = attacker_data.get_defense_chart(&ctx.db)?;
        let attacker = Pokemon::new(attacker_data, attacker_chart, attacker_moves);

        let mut defenders = vec![];

        for defender_name in self.defender_names.iter() {
            let defender_name = ctx.validate::<PokemonRow>(defender_name)?;
            let defender_data = PokemonData::from_name(&defender_name, generation, &ctx.db)?;
            let defender_moves = defender_data.get_moves(&ctx.db)?;
            let defender_chart = defender_data.get_defense_chart(&ctx.db)?;
            let defender = Pokemon::new(defender_data, defender_chart, defender_moves);

            defenders.push(defender);
        }

        for defender in defenders {
            let match_context = MatchComponent {
                defender: &defender,
                attacker: &attacker,
                verbose: self.verbose,
                stab_only: self.stab_only,
            };
            let match_display = DisplayComponent::new(match_context, ctx.config.color_enabled);

            writedoc! {
                writer,
                "
                {match_display}


                "
            }?;
        }

        Ok(0)
    }
}

pub struct CoverageCommand {
    pub names: Vec<String>,
}
impl Command for CoverageCommand {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32> {
        let ctx = DbContext::try_new(config)?;
        let generation = ctx.get_generation()?;

        let mut pokemon = vec![];
        for name in self.names.iter() {
            let name = ctx.validate::<PokemonRow>(name)?;
            let data = PokemonData::from_name(&name, generation, &ctx.db)?;
            let moves = data.get_moves(&ctx.db)?;
            let chart = data.get_defense_chart(&ctx.db)?;

            let mon = Pokemon::new(data, chart, moves);
            pokemon.push(mon);
        }

        let coverage_ctx = CoverageComponent {
            pokemon: &pokemon,
            db: &ctx.db,
        };
        let coverage_display = DisplayComponent::new(coverage_ctx, ctx.config.color_enabled);

        writedoc! {
            writer,
            "
            {coverage_display}
            "
        }?;

        Ok(0)
    }
}

pub struct ResourceCommand {
    pub resource: ResourceArgs,
    pub delimiter: Option<String>,
}
impl Command for ResourceCommand {
    async fn run(&self, config: Config, writer: &mut impl Write) -> Result<i32> {
        let ctx = DbContext::try_new(config)?;
        let delimiter = self.delimiter.clone().unwrap_or("\n".to_string());

        let resource = match self.resource {
            ResourceArgs::Pokemon => PokemonRow::resource(&ctx.db).join(&delimiter),
            ResourceArgs::Moves => MoveRow::resource(&ctx.db).join(&delimiter),
            ResourceArgs::Abilities => AbilityRow::resource(&ctx.db).join(&delimiter),
            ResourceArgs::Types => TypeRow::resource(&ctx.db).join(&delimiter),
            ResourceArgs::Games => GameRow::resource(&ctx.db).join(&delimiter),
        };

        writedoc! {
            writer,
            "
            {resource}
            "
        }?;

        Ok(0)
    }
}

pub struct ConfigCommand {
    pub key: String,
    pub value: Option<String>,
    pub unset: bool,
}
impl Command for ConfigCommand {
    async fn run(&self, _config: Config, writer: &mut impl Write) -> Result<i32> {
        let mut config_file = ConfigFile::from_file()?;
        if self.unset {
            config_file.unset_value(&self.key);
            config_file.save()?;
        } else if let Some(value) = &self.value {
            config_file.set_value(&self.key, value);
            config_file.save()?;
        } else if self.value.is_none() {
            if let Some(value) = config_file.get_value(&self.key) {
                writeln!(writer, "{value}")?;
            }
        }
        Ok(0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resource::config::ConfigBuilder;

    fn config(game: &str) -> Config {
        ConfigBuilder::default()
            .game(String::from(game))
            .color_enabled(false)
            .build()
            .unwrap()
    }

    async fn run_command(command: impl Command, config: Config) -> String {
        let mut writer = vec![];
        command.run(config, &mut writer).await.unwrap();
        String::from_utf8(writer).unwrap()
    }

    #[tokio::test]
    async fn run_pokemon() {
        let config = config("scarlet-violet");
        let pokemon = PokemonCommand {
            name: String::from("ceruledge"),
            moves: false,
            evolution: false,
        };

        let output = run_command(pokemon, config).await;

        insta::with_settings!({
            description => "pokemon ceruledge --game scarlet-violet",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_pokemon_evolution() {
        let config = config("sword-shield");
        let cascoon = PokemonCommand {
            name: String::from("cascoon"),
            moves: false,
            evolution: true,
        };
        let cascoon_output = run_command(cascoon, config.clone()).await;

        insta::with_settings!({
            description => "pokemon cascoon --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(cascoon_output);
        });

        let politoed = PokemonCommand {
            name: String::from("politoed"),
            moves: false,
            evolution: true,
        };
        let politoed_output = run_command(politoed, config.clone()).await;

        insta::with_settings!({
            description => "pokemon politoed --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(politoed_output);
        });

        let applin = PokemonCommand {
            name: String::from("applin"),
            moves: false,
            evolution: true,
        };
        let applin_output = run_command(applin, config.clone()).await;

        insta::with_settings!({
            description => "pokemon applin --evolution --game sword-shield",
            omit_expression => true
        }, {
            insta::assert_snapshot!(applin_output);
        });
    }

    #[tokio::test]
    async fn run_pokemon_moves() {
        let config = config("scarlet-violet");
        let blaziken = PokemonCommand {
            name: String::from("blaziken"),
            moves: true,
            evolution: false,
        };
        let output = run_command(blaziken, config).await;

        insta::with_settings!({
            description => "pokemon blaziken --moves --game scarlet-violet",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_type() {
        let config = config("platinum");
        let ice = TypeCommand {
            primary_type: String::from("ice"),
            secondary_type: None,
        };
        let output = run_command(ice, config.clone()).await;

        insta::with_settings!({
            description => "type ice --game platinum",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });

        let ground_water = TypeCommand {
            primary_type: String::from("ground"),
            secondary_type: Some(String::from("water")),
        };
        let output = run_command(ground_water, config.clone()).await;

        insta::with_settings!({
            description => "type ground water --game platinum",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_move() {
        let config = config("sun-moon");
        let brick_break = MoveCommand {
            name: String::from("brick-break"),
        };
        let output = run_command(brick_break, config).await;

        insta::with_settings!({
            description => "move brick-break --game sun-moon",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_ability() {
        let config = config("black-white");
        let intimidate = AbilityCommand {
            name: String::from("intimidate"),
        };
        let output = run_command(intimidate, config).await;

        insta::with_settings!({
            description => "ability intimidate --game black-white",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }

    #[tokio::test]
    async fn run_match() {
        let config = config("x-y");
        let non_verbose_cmd = MatchCommand {
            defender_names: vec![String::from("golem"), String::from("pachirisu")],
            attacker_name: String::from("lapras"),
            verbose: false,
            stab_only: false,
        };
        let stab_only_cmd = MatchCommand {
            stab_only: true,
            ..non_verbose_cmd.clone()
        };
        let verbose_cmd = MatchCommand {
            verbose: true,
            ..non_verbose_cmd.clone()
        };

        let non_verbose = run_command(non_verbose_cmd, config.clone()).await;
        let stab_only = run_command(stab_only_cmd, config.clone()).await;
        let verbose = run_command(verbose_cmd, config.clone()).await;

        insta::with_settings!({
            description => "match golem pachirisu lapras --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(non_verbose);
        });

        insta::with_settings!({
            description => "match golem pachirisu lapras --stab-only --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(stab_only);
        });

        insta::with_settings!({
            description => "match golem pachirisu lapras --verbose --game x-y",
            omit_expression => true
        }, {
            insta::assert_snapshot!(verbose);
        });
    }

    #[tokio::test]
    async fn run_coverage() {
        let config = config("the-indigo-disk");
        let coverage = CoverageCommand {
            names: vec![
                String::from("flamigo"),
                String::from("cramorant"),
                String::from("ribombee"),
                String::from("ogerpon-cornerstone-mask"),
                String::from("dudunsparce"),
                String::from("sinistcha"),
            ],
        };

        let output = run_command(coverage, config).await;

        insta::with_settings!({
            description => "coverage flamigo cramorant ribombee ogerpon-cornerstone-mask dudunsparce sinistcha --game the-indigo-disk",
            omit_expression => true
        }, {
            insta::assert_snapshot!(output);
        });
    }
}
