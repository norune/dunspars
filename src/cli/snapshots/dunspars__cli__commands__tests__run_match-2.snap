---
source: src/cli/commands.rs
description: match golem pachirisu lapras --stab-only --game x-y
---
golem rock ground
hp    atk   def   satk  sdef  spd   total
80    120   130   55    65    45    495   
lapras water ice
hp    atk   def   satk  sdef  spd   total
130   85    80    85    95    60    535   

lapras's moves vs golem
quad: aqua-tail(p) brine(s) dive(p) hydro-pump(s) surf(s) water-gun(s) water-pulse(s) waterfall(p) whirlpool(s) 
double: avalanche(p) blizzard(s) freeze-dry(s) frost-breath(s) ice-beam(s) ice-shard(p) icy-wind(s) sheer-cold(s) 

golem's moves vs lapras
double: rock-blast(p) rock-slide(p) rock-throw(p) rock-tomb(p) smack-down(p) stone-edge(p) 


pachirisu electric 
hp    atk   def   satk  sdef  spd   total
60    45    70    45    90    95    405   
lapras water ice
hp    atk   def   satk  sdef  spd   total
130   85    80    85    95    60    535   

lapras's moves vs pachirisu
None

pachirisu's moves vs lapras
double: charge-beam(s) discharge(s) electro-ball(s) electroweb(s) nuzzle(p) shock-wave(s) spark(p) thunder(s) thunder-punch(p) thunderbolt(s) volt-switch(s)
